File: leetcode/backtracking/count-numbers-with-unique-digits.py
"""
Given a non-negative integer n, count all numbers with unique digits, x, where 0 ≤ x < 10n.

Example:

Input: 2
Output: 91 
Explanation: The answer should be the total numbers in the range of 0 ≤ x < 100, 
             excluding 11,22,33,44,55,66,77,88,99
"""
# 1st attempt: 5th percentile
# Actually used a backtracking approach
class Solution:
    def countNumbersWithUniqueDigits(self, n: int) -> int:
        if n==0: return 1
        
        def bt(x):
            nonlocal res
            if len(x)==n:
                return
            
            for i in range(0, 10):
                if len(x)==1 and i==0:
                    continue
                
                if i not in visited:
                    res += 1
                    visited.add(i)
                    bt(x + [i])
                    visited.remove(i)
                    
            
            
        visited = set()
        res = 0
        bt([])
        return res

# 2nd attempt: 72nd percentile in speed
# Arithmetic approach
class Solution:
    def countNumbersWithUniqueDigits(self, n: int) -> int:
        def bt(x):
            if x == 0: return 0 
            ans = 9
            for i in range(x-1):
                ans *= (9-i)
            return ans + bt(x-1)
        
        
        return 1 + bt(n)

# 3rd: 72nd percentile/
# Same approach as 2nd attempt, but a bit cleanr
class Solution:
    def countNumbersWithUniqueDigits(self, n: int) -> int:
        if n == 0: return 1
        ans = 9
        for i in range(n-1):
            ans *= (9-i)
        return ans + self.countNumbersWithUniqueDigits(n-1)

File: leetcode/backtracking/count-numbers-with-unique-digits.js
/*
Given a non-negative integer n, count all numbers with unique digits, x, where 0 ≤ x < 10n.

Example:

Input: 2
Output: 91 
Explanation: The answer should be the total numbers in the range of 0 ≤ x < 100, 
             excluding 11,22,33,44,55,66,77,88,99
*/
// 1st attempt: 96th percentile in speed
/**
 * @param {number} n
 * @return {number}
 */
var countNumbersWithUniqueDigits = function(n) {
    var bt = function(x) {
        if (x===0) return 0;
        let ans = 9;
        for (let i = 0; i < x-1; i++) ans *= (9-i);
        return ans + bt(x-1);
    }
    return 1+bt(n);
};

// 3rd attempt: 96th percentile in speed
// I have a habit of adding a new recursive fucntion instead of just making the original stub
// recursive
/**
 * @param {number} n
 * @return {number}
 */
var countNumbersWithUniqueDigits = function(n) {
    if (n===0) return 1;
    let ans = 9;
    for (let i = 0; i < n-1; i++) ans *= (9-i);
    return ans+countNumbersWithUniqueDigits(n-1);
};